### PR TITLE
Add System.ComponentModel.Annotations tests

### LINKED ISSUE #20932: [netfx compat] CompareAttribute throws InvalidOperationException for multi-cased properties
The following code example passes with netfx:
```cs
[Fact]
public static void Validate_LowerAndUpperPropertyName_ThrowsInvalidOperationException()
{
    CompareAttribute attribute = new CompareAttribute(nameof(CompareObject.comparepropertycased));
    Assert.Equal("'CurrentProperty' and 'comparepropertycased' do not match.", attribute.GetValidationResult("b", s_context).ErrorMessage);
    Assert.Equal(ValidationResult.Success, attribute.GetValidationResult(null, s_context));
}
```

In netcoreapp, this throws an InvalidOperationException:
```
System.ComponentModel.DataAnnotations.Tests.CompareAttributeTests.Validate_LowerAndUpperPropertyName_Success [FAIL
  ]
        System.InvalidOperationException : Sequence contains more than one matching element
        Stack Trace:
              at System.Linq.Enumerable.SingleOrDefault[TSource](IEnumerable`1 source, Func`2 predicate)
              at System.ComponentModel.DataAnnotations.CompareAttribute.GetDisplayNameForProperty(Type containerType, S
  tring propertyName)
              at System.ComponentModel.DataAnnotations.CompareAttribute.IsValid(Object value, ValidationContext validat
  ionContext)
              at System.ComponentModel.DataAnnotations.ValidationAttribute.GetValidationResult(Object value, Validation
  Context validationContext)
              at System.ComponentModel.DataAnnotations.Tests.CompareAttributeTests.Validate_LowerAndUpperPropertyName_S
  uccess()
```

### PR DESCRIPTION
**Pull request by hughbe**

### CODE REVIEW

File: src/System.ComponentModel.Annotations/src/Resources/Strings.resx [hunk 1]
```diff
@@ -127,12 +127,18 @@
...
+  </data>
   <data name="MinLengthAttribute_InvalidMinLength" xml:space="preserve">
     <value>MinLengthAttribute must have a Length value that is zero or greater.</value>
   </data>
   <data name="MinLengthAttribute_ValidationError" xml:space="preserve">
     <value>The field {0} must be a string or array type with a minimum length of '{1}'.</value>
   </data>
+  <data name="MinLengthAttribute_InvalidValueType" xml:space="preserve">
+    <value>The field of type {0} must be a string, array or ICollection type.</value>
+  </data>
```
Comment (stephentoub) [inline src/System.ComponentModel.Annotations/src/Resources/Strings.resx]: This is the same string as MaxLengthAttribute_InvalidValueType... why do we need both?

File: src/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/CompareAttribute.cs [hunk 2]
```diff
@@ -56,35 +56,23 @@ protected override ValidationResult IsValid(object value, ValidationContext vali
             {
                 if (OtherPropertyDisplayName == null)
                 {
-                    OtherPropertyDisplayName = GetDisplayNameForProperty(validationContext.ObjectType, OtherProperty);
+                    OtherPropertyDisplayName = GetDisplayNameForProperty(validationContext.ObjectType, otherPropertyInfo);
```
Comment (stephentoub) [inline src/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/CompareAttribute.cs]: Why did this change?

File: src/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/CompareAttribute.cs [hunk 3]
```diff
@@ -56,35 +56,23 @@ protected override ValidationResult IsValid(object value, ValidationContext vali
...
-
             var attributes = CustomAttributeExtensions.GetCustomAttributes(property, true);
             var display = attributes.OfType<DisplayAttribute>().FirstOrDefault();
             if (display != null)
             {
                 return display.GetName();
             }
 
-            return propertyName;
+            return OtherProperty;
```
Comment (stephentoub) [inline src/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/CompareAttribute.cs]: Why did this change?

File: src/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/CustomValidationAttribute.cs [hunk 4]
```diff
@@ -184,12 +184,7 @@ protected override ValidationResult IsValid(object value, ValidationContext vali
             }
             catch (TargetInvocationException tie)
             {
-                if (tie.InnerException != null)
-                {
-                    throw tie.InnerException;
-                }
-
-                throw;
+                throw tie.InnerException;
```
Comment (stephentoub) [inline src/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/CustomValidationAttribute.cs]: Do you know for sure that InnerException will always be non-null in all circumstances?

File: src/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/MaxLengthAttribute.cs [hunk 5]
```diff
@@ -87,9 +87,7 @@ public override bool IsValid(object value)
                 }
                 else
                 {
-                    // A cast exception previously occurred if a non-{string|array} property was passed
-                    // in so preserve this behavior if the value does not implement ICollection
-                    length = ((Array)value).Length;
+                    throw new InvalidCastException(SR.Format(SR.MaxLengthAttribute_InvalidValueType, value.GetType()));
```
Comment (stephentoub) [inline src/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/MaxLengthAttribute.cs]: Why did you change this?

File: src/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/ValidationAttributeStore.cs [hunk 6]
```diff
@@ -212,12 +213,9 @@ internal bool TryGetPropertyStoreItem(string propertyName, out PropertyStoreItem
 
                 if (_propertyStoreItems == null)
                 {
-                    lock (_syncRoot)
+                    lock (_lock)
```
Comment (stephentoub) [inline src/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/ValidationAttributeStore.cs]: As far as I can tell, this is a buggy change.  Why did you change this?  It's now using a process-wide lock to protect instance state, and it removed the double-check that would ensure that multiple threads always see the same value even in the face of race conditions.

File: src/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/Validator.cs [hunk 7]
```diff
@@ -236,7 +237,7 @@ public static void ValidateProperty(object value, ValidationContext validationCo
             var err = GetValidationErrors(value, validationContext, attributes, false).FirstOrDefault();
             if (err != null)
             {
-                err.ThrowValidationException();
+                throw err.GetValidationException();
```
Comment (stephentoub) [inline src/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/Validator.cs]: Why did you change these?  If it's purely about getting to 100% code coverage by allowing the compiler to see that the subsequent brace is unreachable, that's not worth the change.

File: src/System.ComponentModel.Annotations/tests/ValidationExceptionTests.cs [hunk 8]
```diff
@@ -78,5 +82,27 @@ public static void Ctor_String_ValidationAttribute_Object(string errorMessage)
...
+                formatter.Serialize(stream, new SubException());
+
+                stream.Position = 0;
+                Exception ex = Assert.Throws<TargetInvocationException>(() => formatter.Deserialize(stream));
+                Assert.IsType<PlatformNotSupportedException>(ex.InnerException);
+            }
+        }
+
+        [Serializable]
+        public class SubException : ValidationException
```
Comment (stephentoub) [inline src/System.ComponentModel.Annotations/tests/ValidationExceptionTests.cs]: Why is a SubException type necessary for this test?  Shouldn't a PNSE be produced as well when working with ValidationException directly?

File: src/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/CustomValidationAttribute.cs (hunk 4 above)
Comment (hughbe) [inline src/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/CustomValidationAttribute.cs]: This is the `try` block:

```
try
{
    // 1-parameter form is ValidationResult Method(object value)
    // 2-parameter form is ValidationResult Method(object value, ValidationContext context),
    var methodParams = _isSingleArgumentMethod
        ? new object[] { convertedValue }
        : new[] { convertedValue, validationContext };

    var result = (ValidationResult)methodInfo.Invoke(null, methodParams);

    // We capture the message they provide us only in the event of failure,
    // otherwise we use the normal message supplied via the ctor
    _lastMessage = null;

    if (result != null)
    {
        _lastMessage = result.ErrorMessage;
    }

    return result;
}
catch (TargetInvocationException tie)
{
    throw tie.InnerException;
}
```
The only place where the TIE can be thrown is from reflection invoke, and there will always be a non-null inner exception in this case.

File: src/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/ValidationAttributeStore.cs (hunk 6 above)
Comment (hughbe) [inline src/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/ValidationAttributeStore.cs]: Okay. I wasn't sure about the double check

File: src/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/Validator.cs (hunk 7 above)
Comment (hughbe) [inline src/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/Validator.cs]: That was why, yes

File: src/System.ComponentModel.Annotations/tests/ValidationExceptionTests.cs (hunk 8 above)
Comment (hughbe) [inline src/System.ComponentModel.Annotations/tests/ValidationExceptionTests.cs]: ValidationException is not `[Serializable]` so can't be serialized. The test is deserialization, so obviously if we can't serialize, then we can't test deserialization

File: src/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/CompareAttribute.cs (hunk 2 above)
Comment (hughbe) [inline src/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/CompareAttribute.cs]: If you look above, we have the following code:

```
var otherPropertyInfo = validationContext.ObjectType.GetRuntimeProperty(OtherProperty);
```

In the `GetDisplayNameForProperty` method, we used to have the buggy code:

```
var property = containerType.GetRuntimeProperties()
 .SingleOrDefault(
    prop =>
        IsPublic(prop) &&
        string.Equals(propertyName, prop.Name, StringComparison.OrdinalIgnoreCase));
```

However, we already have the property `otherPropertyInfo` above. The name comparison is broken, and `GetRuntimeProperty` only returns public methods so this whole thing is redundant and broken.

So we can just use the actual PropertyInfo we found before

File: src/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/CompareAttribute.cs (hunk 3 above)
Comment (hughbe) [inline src/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/CompareAttribute.cs]: The old value passed to the parameter `propertyName` was `OtherProperty`. Now that that parameter is gone, we use `OtherProperty` directly

File: src/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/MaxLengthAttribute.cs (hunk 5 above)
Comment (lajones) [inline src/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/MaxLengthAttribute.cs]: What if it is an array? You're removing functionality here which was added to solve a previous issue.

---
In reply to: [121382258](https://github.com/dotnet/corefx/pull/20934#discussion_r121382258) [](ancestors = 121382258)

File: src/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/MinLengthAttribute.cs [hunk 9]
```diff
@@ -56,24 +56,19 @@ public override bool IsValid(object value)
...
+                if (value is ICollection collection)
                 {
                     length = collection.Count;
                 }
                 else
                 {
-                    // A cast exception previously occurred if a non-{string|array} property was passed
-                    // in so preserve this behavior if the value does not implement ICollection
-                    length = ((Array)value).Length;
+                    throw new InvalidCastException(SR.Format(SR.LengthAttribute_InvalidValueType, value.GetType()));
```
Comment (lajones) [inline src/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/MinLengthAttribute.cs]: >throw new InvalidCastException(SR.Format(SR.LengthAttribute_InvalidValueType, value.GetType())); [](start = 20, length = 96)

As for MaxLengthAttribute: what if it is an array? You're removing functionality here which was added to solve a previous issue.

File: src/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/MaxLengthAttribute.cs (hunk 5 above)
Comment (lajones) [inline src/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/MaxLengthAttribute.cs]: I think this was addressing [https://github.com/dotnet/corefx/issues/18361](https://github.com/dotnet/corefx/issues/18361)

---
In reply to: [121792346](https://github.com/dotnet/corefx/pull/20934#discussion_r121792346) [](ancestors = 121792346,121382258)

**Comment by lajones** [comment]
@hughbe General comment - could we have a separate PR for each issue please? Would like to address the fix for #20932 separately from the addition of other tests.

File: src/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/MinLengthAttribute.cs (hunk 9 above)
Comment (hughbe) [inline src/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/MinLengthAttribute.cs]: So that's what I first thought when I saw the code and why I fixed this. All Arrays are ICollections, so this code path will never be hit as they will take the precious if block. We have tests that cover this in the project.
What this is actually doing is simulating an InvalidCastException for comparability with before we added ICollection support.

**Comment by hughbe** [comment]
Gladly! Will do so soon

File: src/System.ComponentModel.Annotations/tests/ValidationExceptionTests.cs (hunk 8 above)
Comment (hughbe) [inline src/System.ComponentModel.Annotations/tests/ValidationExceptionTests.cs]: I've deleted this test

**Comment by hughbe** [comment]
I've extracted the netfx compat fix to #21031, this just adds some tests and cleans up some code

**Comment by hughbe** [comment]
Test Portable Linux x64 Debug Build

**Comment by hughbe** [comment]
I rebased this PR to fix merge conflicts

File: src/System.ComponentModel.Annotations/tests/CustomValidationAttributeTests.cs [hunk 10]
```diff
@@ -69,6 +69,26 @@ public static void Constructor(Type validatorType, string method)
             Assert.Equal(method, attribute.Method);
         }
 
+        [Fact]
+        public void FormatErrorMessage_NoError_ContainsName()
```
Comment (lajones) [inline src/System.ComponentModel.Annotations/tests/CustomValidationAttributeTests.cs]: Nit: Just checking - did you mean this to say "NoError"? The test is about what happens to the error message? The test below checks that validation fails - so did you mean "validation passes" here? If so maybe add that assert or rename the tests.

Comment (hughbe) [inline src/System.ComponentModel.Annotations/tests/CustomValidationAttributeTests.cs]: It's not quite that validation passes. It's actually that no validation has ever been performed on that instance - i've updated the comments accordingly

File: src/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/MinLengthAttribute.cs (hunk 9 above)
Comment (lajones) [inline src/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/MinLengthAttribute.cs]: That makes sense - thanks for clarifying.

File: src/System.ComponentModel.Annotations/tests/CustomValidationAttributeTests.cs (hunk 10 above)
Comment (lajones) [inline src/System.ComponentModel.Annotations/tests/CustomValidationAttributeTests.cs]: OK. Now it all makes sense. Thanks for the updates.

**Comment by lajones** [comment]
@ajcvickers This looks goods to me. Do you want to review before @hughbe merges this in?

**Comment by hughbe** [comment]
I don't actually have permission to merge it in as I'm not an MSFT employee so that's up to you guys :) thanks for reviewing

**Comment by lajones** [comment]
My bad. I meant "before we merge this in for @hughbe" 😄

**Comment by hughbe** [comment]
Friendly ping :)
